[Ready for Review] Update BinderHub workshop

sgibson91 opened the pull request:
<!--
Please complete the following sections when you submit your pull request. You are encouraged to keep this top level comment box updated as you develop and respond to reviews. Note that text within html comment tags will not be rendered.
-->
### Summary

<!-- Describe the problem you're trying to fix in this pull request. Please reference any related issue and use fixes/close to automatically close them, if pertinent. For example: "Fixes #58", or "Addresses (but does not close) #238". -->

This PR is a reconstruction of #762 

Addresses #694 

This PR is implementing improvements to the Zero to BinderHub workshop recommended through feedback from RSE Con UK 2019.

### List of changes proposed in this PR (pull-request)

<!-- We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

- See check list in issue #694 for changes to the actual workshop.
- Helm section has been made Helm3 compatible
- A hackmd template has been added for collecting feedback/distributing links during the workshop

### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

- [x] Everything looks ok?

### Acknowledging contributors

<!-- Please select the correct box -->

- [x] All contributors to this pull request are already named in the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file.
- [x] The following people should be added to the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file: <!-- replace this text with the GitHub IDs of any new contributors -->

sgibson91 commented on workshops/build-a-binderhub/workshop-presentations/zero-to-binderhub.md:
```diff
@@ -197,43 +198,63 @@ Do this **before** adding any secrets to the folder!
+A short (but by no means exhaustive) [glossary](#glossary-of-kubernetes-terms) of Kubernetes terms is given at the end of this workshop, should you require further explanation.
 
-**NOTE:** You shouldn't need to login to Azure when using the Cloud Shell as it reads your login from the Portal.
+#### 1. Login to your account
```
This is wrong ! Please change!

sgibson91 commented on workshops/build-a-binderhub/workshop-presentations/zero-to-binderhub.md:
````diff
@@ -197,43 +198,63 @@ Do this **before** adding any secrets to the folder!
 
-To see a list of Azure subscriptions you have available to you, you can run the following command.
+```bash
+az login
````
```suggestion
az log-in
```

malvikasharan commented:
Not sure who should look at this one. Gonna tag @betatim if he can suggest someone who can review this.

sgibson91 commented:
@consideRatio might be able to help?

Lovkush-A commented on workshops/build-a-binderhub/binderhub_resources/collaborative_notes_template.md:
```diff
@@ -0,0 +1,79 @@
+[comment]: <This is a templatefor taking collaborative notes during a Build a BinderHub workshop. It has been formatted to be Markdown compatible and should "just work" when copy-pasted into HackMD.io>
```
Typo: 'templatefor'

sgibson91 replied:
```suggestion
[comment]: <This is a template for taking collaborative notes during a Build a BinderHub workshop. It has been formatted to be Markdown compatible and should "just work" when copy-pasted into HackMD.io>
```

Lovkush-A commented on workshops/build-a-binderhub/binderhub_resources/collaborative_notes_template.md:
```diff
@@ -0,0 +1,79 @@
+
+[comment]: <It's useful to create a short bit.ly link for the hackmd pad>
+* This HackMD pad:
+* workshop instructions: [bit.ly/zero-to-binderhub-workshop](https://bit.ly/zero-to-binderhub-workshop)
```
Typo: 'w' should be capitalised?

sgibson91 replied:
```suggestion
* Workshop instructions: [bit.ly/zero-to-binderhub-workshop](https://bit.ly/zero-to-binderhub-workshop)
```

Lovkush-A commented on workshops/build-a-binderhub/binderhub_resources/collaborative_notes_template.md:
```diff
@@ -0,0 +1,79 @@
+[comment]: <If there are slides accompanying the workshop, link to them here>
+* The intro slides:
+* The Turing Way repo: [alan-turing-institute/the-turing-way](https://github.com/alan-turing-institute/the-turing-way)
+* Chat with us on Gitter!
```
Include a link to Turing Way slack channel?

sgibson91 replied:
```suggestion
* Chat with us on [Slack](https://join.slack.com/t/theturingway/shared_invite/zt-fn608gvb-h_ZSpoA29cCdUwR~TIqpBw)!
```

sgibson91 commented:
Thanks @Lovkush-A, I've addressed all your points! 🙌🏻

sgibson91 commented:
If this PR is not getting reviewed/approved because folk feel like they don't know enough about K8s and/or helm, all that _really_ needs to be reviewed is whether the instructions in `zero-to-binderhub.md` follow along with/make sense compared to the [Azure instructions in the Zero to JupyterHub docs](https://zero-to-jupyterhub.readthedocs.io/en/stable/kubernetes/index.html), as this is basically just an Azure-specific, one-pager copy of those docs.

Lovkush-A commented on workshops/build-a-binderhub/workshop-presentations/zero-to-binderhub.md:
```diff
@@ -157,7 +158,7 @@ This will make the BinderHub setup less intensive.
 
 Make a folder to store this files and change into it.
```
typo: 'this' should be 'the' or 'these'.

sgibson91 replied:
```suggestion
Make a folder to store these files and change into it.
```

Lovkush-A commented on workshops/build-a-binderhub/workshop-presentations/zero-to-binderhub.md:
```diff
@@ -176,15 +177,15 @@ wget -O secret-template.yaml http://bit.ly/secret-template
 
 Make the shell script executable will the following command.
```
typo: 'will' should be 'with'

sgibson91 replied:
```suggestion
Make the shell script executable with the following command.
```

Lovkush-A commented on workshops/build-a-binderhub/workshop-presentations/zero-to-binderhub.md:
````diff
@@ -197,43 +198,63 @@ Do this **before** adding any secrets to the folder!
     --output table
 ```
+
 * `--name` specifies the name of your resource group and should be something that uniquely identifies this hub.
````
Potential source of confusion? "uniquely identifies this hub".  Multiple people in a workshop will be using the same name `testhub` which suggests you do not need to have a unique name. I do not know anything about Azure so I am not sure. A possible re-wording:
"...name of your resource group. In practice this should be something that uniquely identifies this hub, but for this workshop, we will all use the same name `testhub`"

sgibson91 replied:
```suggestion
* `--name` specifies the name of your resource group and should be something that uniquely identifies this hub from other hubs you may deploy.
```

Lovkush-A commented on workshops/build-a-binderhub/workshop-presentations/zero-to-binderhub.md:
```diff
@@ -242,24 +263,26 @@ az group create --name testhub \
+#### 4. Create an Azure Kubernetes Service (AKS) Cluster
 
 This command will request a Kubernetes cluster within the resource group we created.
 It will request one `Standard_D2s_v3` virtual machine which a Kubernetes cluster installed.
```
typo, I think: 'which' should be 'with'

sgibson91 replied:
```suggestion
It will request a Kubernetes cluster with one `Standard_D2s_v3` virtual machine as a 'node'.
```

Lovkush-A commented on workshops/build-a-binderhub/workshop-presentations/zero-to-binderhub.md:
```diff
@@ -275,6 +298,32 @@ az aks create --name hubcluster \
+
+#### A. Locate your Resource Groups
+
+In Azure Portal, select "Resource Groups from the left-most menu panel.
```
typo: missing quotation mark

sgibson91 replied:
```suggestion
In Azure Portal, select "Resource Groups" from the left-most menu panel.
```

Lovkush-A commented on workshops/build-a-binderhub/workshop-presentations/zero-to-binderhub.md:
```diff
@@ -275,6 +298,32 @@ az aks create --name hubcluster \
+
+#### D. View the logs for Managed Cluster
+
+The logs for deploying the cluster will be under "Managed Cluster".
```
Suggestion: say something about what they are looking for in the logs. E.g. that Status says "Suceeded".  If you do decide to make this change, then maybe change the accompanying image to have box around the whole "Managed Cluster"'s row, or have arrows pointing to "Managed Cluster" and to "Suceeded"

Lovkush-A commented on workshops/build-a-binderhub/workshop-presentations/zero-to-binderhub.md:
```diff
@@ -314,100 +365,39 @@ aks-nodepool1-97000712-0   Ready    agent   19m   v1.9.11
-:question: :question: :question: :question: :question:
-
-#### 3. Initialise `helm` and `tiller`
+#### 1. Verify Helm
```
In jupyterhub resource, they have installation step before verification.

sgibson91 replied:
Yes, they do. I normally run this workshop using the Azure Cloud Shell however, which has a lot of tools preinstalled so I don't have to worry about getting people setup before the day.

Lovkush-A commented on workshops/build-a-binderhub/workshop-presentations/zero-to-binderhub.md:
```diff
@@ -457,19 +448,22 @@ helm repo update
   Available versions can be found [here](https://jupyterhub.github.io/helm-chart/#development-releases-binderhub).
   We have used the version released on 11 August 2019.
-* `--name` and `--namespace` may be different, but it's recommended they be the same to avoid confusion.
+* It is recommended that `--namespace` be the same as the provided `NAME` in order to avoid confusion.
```
Where is `NAME` provided? Possible I just missed it.

sgibson91 replied:
`NAME` is `binderhub`. The Helm CLI has updated so NAME is now a positional argument instead of an optional arg with a flag.

Lovkush-A commented:
@sgibson91 I tried having a look and comparing with the resources on jupyterhub and binderhub. It looks like they match up.

sgibson91 commented on workshops/build-a-binderhub/workshop-presentations/zero-to-binderhub.md:
```diff
@@ -457,19 +448,22 @@ helm repo update
   Available versions can be found [here](https://jupyterhub.github.io/helm-chart/#development-releases-binderhub).
   We have used the version released on 11 August 2019.
-* `--name` and `--namespace` may be different, but it's recommended they be the same to avoid confusion.
+* It is recommended that `--namespace` be the same as the provided `NAME` in order to avoid confusion.
```
```suggestion
* It is recommended that `--namespace` be the same as the provided `NAME` in order to avoid confusion.
  In this case, our `NAME` is `binderhub`.
```

sgibson91 commented:
Hey @Lovkush-A, I think I have addressed all your comments. If you're happy, can you approve this Pull Request following the steps [here](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/approving-a-pull-request-with-required-reviews) please? (But select "Approve", not "Request Changes" like in the screenshot!) Then the PR will be mergeable. Thank you! 👾

Lovkush-A reviewed (approved):
All changes look good!

Also worth saying that tutorial looks in top shape; very clear and concise. Despite me having no idea about most of the content, I was able to follow along in the tutorial and make sense of the instructions!

sgibson91 commented:
Thank you @Lovkush-A, glad to know it's a useful resource!
